File: src/Tree/BinaryTreeBFSGuide/PopulatingNextRightPointersinEachNode/PopulatingNextRightPointersinEachNode.java
class Solution {
    public Node connect(Node root) {
        if(root == null) return root;
        
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        
        
        while(!q.isEmpty()){
            int size = q.size();
            //Create a dummy node 
            Node pre = new Node();
            for(int i = 0; i < size; i++){
                Node top = q.poll();
                pre.next = top;
                pre = top;
                //Add the next level nodes onto the queue
                if(top.left != null){
                    q.add(top.left);
                }
                if(top.right != null){
                    q.add(top.right);
                }
            }
        }
        return root;
    }
}

File: src/Tree/BinaryTreeBFSGuide/PopulatingNextRightPointersinEachNode/PopulatingNextRightPointersinEachNode.py
from collections import deque

class Solution:
    def connect(self, root: 'Node') -> 'Node':
    
        if root is None:
            return root
        
        queue = deque([root])
        
        while len(queue) != 0:
            size = len(queue)
            
            dummy = Node()
            pre = dummy
            
            for idx in range(size):
                cur = queue.popleft()
                pre.next = cur
                pre = cur
                
                if cur.left is not None:
                    queue.append(cur.left)
                if cur.right is not None:
                    queue.append(cur.right)
        
        return root
